text/0193-removal-of-type-unions.md

```diff
@@ -0,0 +1,575 @@
+
+Alternatives were considered on both sides of the problem:
+
+- [Removing ambiguity in the `@jsii/kernel`](#removing-ambiguity-in-the-`@jsii/kernel`)
+- [Language-specific union modeling](#language-specific-union-modeling)
+
+### Removing ambiguity in the `@jsii/kernel`
+
+#### Dynamic disambiguation
+
+Since the `@jsii/kernel` has access to the type model, it has information about the candidate types including what their
+members are. Matching the members of the object instance against the candidate types can often determine the correct
+dynamic type of the value.
+
+This kind of matching can however be expensive (especially if _deep_ matching is necessary), and cannot remove all
+ambiguity: **JavaScript** provides no runtime-available information about the signature of a method (making it
+impossible to distinguish two methods with the same name but different signatures), and types that differ only in
+optional properties are impossible to tell apart. This would likely require restricting _type unions_ so that instances
+can be matched to exactly one of the candidates conclusively, for example by requiring that each type in a _type union_
+has at least one member that is not present in other candidates.
+
+#### Explicit disambiguation
+
+One solution to remove the ambiguity is to require developers manually annotate object literals returned through a _type
+union_ with the actual type information for the value they are returning, using a helper library provided by `jsii`:
+
+```ts
+import { typed } from '@jsii/types';
+// ...
+export class Example {
+  // ...
+  public fooOrBar(): Foo | Bar {
+    return typed({ foo: true }, 'example.Foo');
+  }
+  // ...
+}
+```
+
+This would require specific machinery to be built into the `jsii` compiler to validate when this annotation is required
+and fail compilation otherwise; or accepting that non-**TypeScript** (nor **Javascript**) users would receive a runtime
+error if this was not done when it was required.
+
+The specific syntax for this explicit annotation could also be awkward to come up with: _interfaces_ have no
+representation in **Javascript**, and an _interface_ name cannot be passed as an argument like a _class_ name would.
+This means users would have to manually type the _jsii fully qualified name_ of the type by hand. Alternatively, a
+sophisticated **TypeScript** _transformation_ would need to be added to the `jsii` compiler to transparently process the
+annotation call, replacing an interface name with the appropriate _jsii fully qualified name_.
+
+#### API Extension
+
+The `@jsii/kernel` API could be extended so that instances can be tagged using unions, instead of necessarily narrowing
+down to a single type. This way, the kernel would not have to make any decision beyond the statically available context
+of a serialization.
+
+This requires more extensive work to be performed in each individual language runtime library.
+
+### Language-specific union modeling
```

Comment:
Is this needed here really? Feels like a future RFC for *adding support* for type unions, no?

Reply:
Alternatives to removing type unions include **not** removing type unions, and only fixing them up so the current problems no longer exist.

Reply:
Okay